Add Topology sidebar for Operator Backed Services

rottencandy opened the pull request:
**Fixes**: 
https://issues.redhat.com/browse/ODC-4024
<!-- For e.g Fixes: https://issues.redhat.com/browse/ODC-XXX -->

**Analysis / Root cause**: 
As a user, I want to see more details about operator backed services in Topology.
<!-- Briefly describe analysis of US/Task or root cause of Defect -->

**Solution Description**: 
Add a sidebar that can be opened by clicking an operator backed service from the topology.
<!-- Describe your code changes in detail and explain the solution -->

**Screen shots / Gifs for design review(updated)**: 
![p1](https://user-images.githubusercontent.com/20013884/88057468-c2572e80-cb7f-11ea-839c-8270f73ca4b0.png)
![p2](https://user-images.githubusercontent.com/20013884/88057483-c71be280-cb7f-11ea-98e2-f7a45378f351.png)


cc: @openshift/team-devconsole-ux @rohitkrai03 

<!-- If change affects UI in any way, tag @openshift/team-devconsole-ux and add screenshots/gifs  -->

**Browser conformance**: 
<!-- To mark tested browsers, use [x] -->
- [ ] Chrome
- [x] Firefox
- [ ] Safari
- [ ] Edge

invincibleJai commented on frontend/packages/dev-console/src/components/topology/operators/TopologyOperatorBackedPanel.tsx:
```diff
@@ -0,0 +1,45 @@
+import TopologyOperatorBackedResources from './TopologyOperatorBackedResources';
+import { TopologyDataObject } from '../topology-types';
+
+export type TopologyResourcePanelProps = {
```
```suggestion
export type TopologyOperatorBackedPanelProps = {
```
nit: can rename props based on component name

rottencandy replied:
Updated.

invincibleJai commented on frontend/packages/dev-console/src/components/topology/operators/TopologyOperatorBackedPanel.tsx:
```diff
@@ -0,0 +1,45 @@
+            <ResourceLink
+              kind={referenceFor(obj)}
+              name={name}
+              namespace={getActiveNamespace()}
```
why not pick `namespace` from obj itself?

rottencandy replied:
Updated.

invincibleJai commented on frontend/packages/dev-console/src/components/topology/operators/TopologyOperatorBackedPanel.tsx:
```diff
@@ -0,0 +1,45 @@
+      <div className="overview__sidebar-pane-head resource-overview__heading">
+        <h1 className="co-m-pane__heading">
+          <div className="co-m-pane__name co-resource-item">
+            <ResourceIcon kind="Operator" />
```
can use model to get kind or pick from obj itself, wdyt?

rottencandy replied:
I added a separate icon because `ResourceLink` was showing a `CSV` icon, but design docs had `O`, and `O` is also consistent with topology node. I can change it if it's not required.

invincibleJai commented on frontend/packages/dev-console/src/components/topology/operators/TopologyOperatorBackedResources.tsx:
```diff
@@ -0,0 +1,41 @@
+const TopologyOperatorBackedResources: React.FC<TopologyOperatorBackedResourcesProps> = ({
+  item,
+}) => {
+  const combinedResources = item.groupResources.reduce<OverviewItem>(
```
will `groupResources` be always defined as `groupResources` is optional param in `TopologyDataObject` ? so may be good to have a check

rottencandy replied:
Added default `[]` value during destructure.

invincibleJai commented on frontend/packages/dev-console/src/components/topology/operators/TopologyOperatorBackedResources.tsx:
```diff
@@ -0,0 +1,41 @@
+    { buildConfigs: [], pods: [], obj: {}, routes: [], services: [] },
+  );
+  const { buildConfigs, pods, obj, routes, services } = combinedResources;
+  combinedResources.isOperatorBackedService = true;
```
can we have it as default value is accumulator for reducer? WDYT

rottencandy replied:
Updated.

invincibleJai commented on frontend/packages/dev-console/src/components/topology/operators/TopologyOperatorBackedResources.tsx:
```diff
@@ -0,0 +1,41 @@
+}) => {
+  const combinedResources = item.groupResources.reduce<OverviewItem>(
+    (acc, val) => {
+      _.keys(acc).forEach((key) => {
```
i know we have mix of lodash and core utils but having in same line , maybe can use `Object.keys`

rottencandy replied:
Updated. No longer using `keys` function.

invincibleJai commented:
on view all pods , it shows just 1 but in sidebar says 5

![Jul-17-2020 15-52-22](https://user-images.githubusercontent.com/5129024/87776648-b993ef00-c845-11ea-9593-e31e3a96806e.gif)

rottencandy commented:
Rebased and updated with `Details` tab.

rohitkrai03 commented on frontend/packages/dev-console/src/components/topology/operators/TopologyOperatorBackedResources.tsx:
```diff
@@ -0,0 +1,35 @@
+  return (
+    <div className="overview__sidebar-pane-body">
+      <OperatorBackedOwnerReferences item={ownerReferencedResources as OverviewItem} />
+      <TopologyHelmReleaseResourcesPanel
+        manifestResources={finalRes}
+        releaseNamespace={item.resources.obj.metadata.namespace}
+      />
```
Probably better to rename this component to something generic that both operator and helm side panels use. Something like `TopologyWorkloadResourcesPanel`. WDYT @jeff-phillips-18 ?

jeff-phillips-18 replied:
agreed.

beaumorley commented:
Looks good @rohitkrai03 @rottencandy 

One small comment. In the second screen capture with the resources tab selected, the alignment of the resources is off.

It looks like this:
![image](https://user-images.githubusercontent.com/52013984/88210377-187cae00-cc22-11ea-9ba5-9ec2077e8479.png)

The "Deployments" title should align with the "D" in the details tab. The deployments should move over the same amount of space (but stay indented from the title like they are now).

This looks like it may be resolved in the third screen grab taken by Jai but want to make sure. 

Thank you!

cc @serenamarie125

rottencandy commented:
The extra space was occuring due to duplicate `div`s in the resources component, fixed it.

I also renamed `TopologyHelmReleaseResourcesPanel` to `TopologyGroupResourcesPanel` and moved it from `helm` to the `components` directory.

cc: @beaumorley  @rohitkrai03 
![demo](https://user-images.githubusercontent.com/20013884/88323843-78697680-cd40-11ea-9ae7-1bcf4a0c63ac.png)

debsmita1 commented on frontend/packages/dev-console/src/components/topology/components/TopologyGroupResourcesPanel.tsx:
```diff
@@ -22,24 +22,19 @@ const TopologyHelmReleaseResourcesPanel: React.SFC<TopologyHelmReleaseResourcesP
     return lists;
   }, []);
-
-  return <div className="overview__sidebar-pane-body">{resourceLists}</div>;
```
removing the div caused this in the Helm group's sidepanel
![helm-sidepanel](https://user-images.githubusercontent.com/22490998/88334185-03ea0400-cd4f-11ea-8721-0a31c75f8162.png)

rottencandy replied:
Had wrapped the wrong component. Updated.
![demo](https://user-images.githubusercontent.com/20013884/88363172-b5fbed00-cd9c-11ea-89ce-6162fcae7a81.png)

debsmita1 commented:
/lgtm

rohitkrai03 commented:
/assign

rohitkrai03 commented on frontend/packages/dev-console/src/components/topology/operators/TopologyOperatorBackedPanel.tsx:
```diff
@@ -0,0 +1,58 @@
+
+  const ResourcesSection = () => <TopologyOperatorBackedResources item={item} />;
+  const DetailsSection = () => (
+    <div className="co-m-pane__body">
```
This will align the resource details better with heading and nav items of the sidebar. There's a difference of `10px` between sidebar heading and resource details.
```suggestion
    <div className="overview__sidebar-pane-body">
```

rottencandy replied:
Updated.
